### PR TITLE
Generalizes halo filling for flux boundary conditions

### PR DESCRIPTION
**Pull request by glwagner**
This PR generalizes `fill_halo_regions!` for flux boundary conditions to work correctly for biharmonic diffusivities, in addition to second-order diffusivities. This will allow biharmonic diffusivities to be used in closed domains or along `Bounded` dimensions.

In particular (and after tweaking the definition of `AnisotropicBiharmonicDiffusivity`), no flux conditions on any even diffusion operator are enforced by ensuring that odd derivatives across the boundary are zero. This was already done for the first derivative, but did not work for third derivatives and higher. In this PR, kernels are written that _reflect_ fields across the boundary for flux boundary conditions. Reflection implies that odd derivatives are zero and ensures that no flux is enforced for diffusivities of order higher than two.

Note that `Value` and `Gradient` boundary conditions do not work for biharmonic diffusivity (and likely never will). Thus only `Flux` and `Periodic` boundary conditions are supported with biharmonic diffusivity.

This PR also does some minor shuffling of file content and names, and generalizes the tracer and momentum budget tests.

### CODE REVIEW

File: src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_biharmonic_diffusivity.jl [hunk 1]
```diff
@@ -3,64 +3,87 @@
...
+`κx`, `κy`, and `κz` may be `NamedTuple`s with fields corresponding
 to each tracer, or a single number to be a applied to all tracers.
+
+If `νh` or `κh` are provided, then `νx = νy = νh` or `κx = κy = κh`.
+
 The tracer flux divergence associated with an anisotropic biharmonic diffusivity is, for example
 
 ```math
-    ∂ᵢ κᵢⱼ ∂ⱼc = (κh ∇⁴h + κz ∂⁴z) c
+    ∂ᵢ κᵢⱼ ∂ⱼc = - (κh ∂⁴x + ∂⁴y + κz ∂⁴z) c
```
Comment (ali-ramadhan) [inline src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_biharmonic_diffusivity.jl:29]: Should probably be
```
∂ᵢ κᵢⱼ ∂ⱼ c = - (κx ∂⁴x + κy ∂⁴y + κz ∂⁴z) c
```

**Review by ali-ramadhan — CHANGES_REQUESTED** [review]
> In particular (and after tweaking the definition of `AnisotropicBiharmonicDiffusivity`), no flux conditions on any even diffusion operator are enforced by ensuring that odd derivatives across the boundary are zero. This was already done for the first derivative, but did not work for third derivatives and higher. In this PR, kernels are written that reflect fields across the boundary for flux boundary conditions. Reflection implies that odd derivatives are zero and ensures that no flux is enforced for diffusivities of order higher than two.

Am I right in understanding that now `_fill_*_halo!(c, ::FBC, H, N)` is called for all flux boundary conditions but is only required if running with biharmonic diffusivity (or other higher-order diffusivity closures)?

If so, it seems that the vast majority of models will be launching extra kernels (small performance hit) if they have flux boundary conditions. We should open an issue about fixing this potential performance hit.

> Note that `Value` and `Gradient` boundary conditions do not work for biharmonic diffusivity (and likely never will). Thus only `Flux` and `Periodic` boundary conditions are supported with biharmonic diffusivity.

Should we add a warning somewhere in case a user decides to use biharmonic diffusivity with `Value` or `Gradient` boundary conditions, or is biharmonic diffusivity still an experimental feature?

---

Couple of tests failing due to `type AnisotropicBiharmonicDiffusivity has no field κh`: https://travis-ci.com/github/CliMA/Oceananigans.jl/jobs/380678166#L509-L515

And the Eady turbulence example needs to be fixed.

**Comment by glwagner** [comment]
We have always launched a kernel for flux bcs! Previously we used broadcasting. When the halo size was greater than 1, the intent of the broadcast operation was violated. This PR corrects that.

Some informal benchmarks suggest that a KernelAbstractions kernel  can be more efficient than using GPUArrays broadcasting. We can also use multi threading. Thus I would expect speed up from this change, rather than slowdown (but likely negligible speed up). But of course benchmarking is a good idea.

Another advantage is that we can utilize KernelAbstractions dependency feature more fully if we get to the point where we do not use broadcasting for anything. We haven't optimized our dependency structure, however. Also, there are a few puzzles to be worked out regarding periodic boundary conditions before we can convert the periodic halo filling to kernels.

**Review by ali-ramadhan — APPROVED** [review]
Ah I see, right. I agree more kernels would be nice here. Don't think broadcasting is multithreaded?

> Note that `Value` and `Gradient` boundary conditions do not work for biharmonic diffusivity (and likely never will).

I think we should mention this to users somewhere (warning and/or docstring, or documentation).

Otherwise looks good to merge once tests pass.

**Comment by glwagner** [comment]
> I think we should mention this to users somewhere (warning and/or docstring, or documentation).

I agree. I should clarify too. One reason that biharmonic is more complicated is we need more than one boundary condition. This PR actually modifies the `Flux` boundary condition so that it applies multiple boundary conditions (the same number as the number of halos) in order to enforce zero diffusive flux (for the diffusion operators we support --- our method is not general mathematically). With `Value` and `Gradient` we would need the user to provide another boundary condition to fully specify the problem. The one situation where this might make sense is in the case of an open boundary condition with non zero `NormalFlow`. In that case one might wish to specify a `Value` and then perhaps also another condition. It's worth some thought and perhaps reading in the literature...

For now, I will implement an error.

File: src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_biharmonic_diffusivity.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (ali-ramadhan) [reply src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_biharmonic_diffusivity.jl:29]: You ignored my suggestion :(

[the remark this answers is not in the document]
Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_biharmonic_diffusivity.jl:29]: bahhh sorry about that
